core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/util/SignerUtils.java

```diff
@@ -224,22 +151,80 @@ private static byte[] sign(byte[] data, byte[] key, SigningAlgorithm algorithm)
+     */
     public static <T> T validatedProperty(SignRequest<?, ?> request, SignerProperty<T> property) {
-        return Validate.notNull(request.property(property), property.toString() + "must not be null!");
+        return Validate.notNull(request.property(property), property.toString() + " must not be null!");
     }
 
+    /**
+     * Validate that the {@link SignerProperty} is present in the {@link SignRequest}.
+     * <p>
+     * The value, {@link T}, is return when present, and the default is returned otherwise.
+     */
     public static <T> T validatedProperty(SignRequest<?, ?> request, SignerProperty<T> property, T defaultValue) {
         return Validate.getOrDefault(request.property(property), () -> defaultValue);
     }
+
+    /**
+     * Add the host header based on parameters of a request
+     */
+    public static void addHostHeader(SdkHttpRequest.Builder requestBuilder) {
+        // AWS4 requires that we sign the Host header, so we
+        // have to have it in the request by the time we sign.
+
+        StringBuilder hostHeaderBuilder = new StringBuilder(requestBuilder.host());
+        if (!SdkHttpUtils.isUsingStandardPort(requestBuilder.protocol(), requestBuilder.port())) {
+            hostHeaderBuilder.append(":").append(requestBuilder.port());
+        }
+
+        requestBuilder.putHeader(SignerConstant.HOST, hostHeaderBuilder.toString());
+    }
+
+    /**
+     * Add a date header using a datetime string
+     */
+    public static void addDateHeader(SdkHttpRequest.Builder requestBuilder, String dateTime) {
+        requestBuilder.putHeader(SignerConstant.X_AMZ_DATE, dateTime);
+    }
+
+    public static void addSha256ContentHeader(SdkHttpRequest.Builder requestBuilder, ContentChecksum contentChecksum) {
+        requestBuilder.firstMatchingHeader(SignerConstant.X_AMZ_CONTENT_SHA256)
+            .filter(h -> h.equals("required"))
+            .ifPresent(h ->
+                requestBuilder.putHeader(
+                    SignerConstant.X_AMZ_CONTENT_SHA256, contentChecksum.contentHash()));
+    }
+
+    /**
+     * Add a checksum header if the checksum is not null and the payload is signed
+     */
+    public static void addChecksumHeader(SdkHttpRequest.Builder requestBuilder, SdkChecksum sdkChecksum,
+                                         String contentHashString,
+                                         String checksumHeaderName) {
+
+        if (sdkChecksum != null && !isUnsignedPayload(contentHashString)) {
+            requestBuilder.putHeader(checksumHeaderName, BinaryUtils.toBase64(sdkChecksum.getChecksumBytes()));
+        }
+    }
+
+    /**
+     * Check if a payload is unsigned based on the content hash
+     */
+    private static boolean isUnsignedPayload(String contentHashString) {
+        return "UNSIGNED-PAYLOAD".equals(contentHashString) || "STREAMING-UNSIGNED-PAYLOAD-TRAILER".equals(contentHashString);
```

Comment:
I see you fixed from `_` to `-`. Are there tests for both these cases? Or atleast make sure they are covered as part of other signers later.

Reply:
Other signers.